linera-views/src/views/key_value_store_view.rs

```diff
@@ -1164,45 +1115,1304 @@ where
+    /// view.insert(vec![3, 4], vec![42]).unwrap();
+    /// let key_values = view.find_key_values_by_prefix(&[0]);
+    /// assert_eq!(key_values, vec![(vec![1], vec![34])]);
+    /// # })
+    /// ```
+    pub fn find_key_values_by_prefix(&self, key_prefix: &[u8]) -> Vec<(Vec<u8>, Vec<u8>)> {
+        let mut key_values = Vec::new();
+        let len_prefix = key_prefix.len();
+        for (key, value) in self.state.range(get_interval(key_prefix.to_vec())) {
+            let key = key[len_prefix..].to_vec();
+            key_values.push((key, value.clone()));
+        }
+        key_values
+    }
+
+    fn compute_hash(&self) -> Result<<sha3::Sha3_256 as Hasher>::Output, ViewError> {
+        let mut hasher = sha3::Sha3_256::default();
+        let mut count = 0u32;
+        self.for_each_index_value(|index, value| -> Result<(), ViewError> {
+            count += 1;
+            hasher.update_with_bytes(index)?;
+            hasher.update_with_bytes(value)?;
+            Ok(())
+        })?;
+        hasher.update_with_bcs_bytes(&count)?;
+        Ok(hasher.finalize())
+    }
+}
+
+impl<C> HashableView<C> for SmallKeyValueStoreView<C>
+where
+    C: Context + Send + Sync,
+    ViewError: From<C::Error>,
+{
+    type Hasher = sha3::Sha3_256;
+
+    async fn hash_mut(&mut self) -> Result<<Self::Hasher as Hasher>::Output, ViewError> {
+        self.compute_hash()
+    }
+
+    async fn hash(&self) -> Result<<Self::Hasher as Hasher>::Output, ViewError> {
+        self.compute_hash()
+    }
+}
+
+/// A combined store.
+#[derive(Debug)]
+pub enum KeyValueStoreView<C> {
+    /// The case of a small key value store
+    Small(SmallKeyValueStoreView<C>),
+    /// The case of a big key value store
+    Big(BigKeyValueStoreView<C>),
+}
+
+impl<C> View<C> for KeyValueStoreView<C>
+where
+    C: Context + Send + Sync,
+    ViewError: From<C::Error>,
+{
+    const NUM_INIT_KEYS: usize = NUM_INIT_KEYS;
+
+    fn context(&self) -> &C {
+        match self {
+            Self::Small(small) => small.context(),
+            Self::Big(big) => big.context(),
+        }
+    }
+
+    fn pre_load(context: &C) -> Result<Vec<Vec<u8>>, ViewError> {
+        let key_total_size = context.base_key().base_tag(KeyTag::TotalSize as u8);
+        Ok(vec![key_total_size])
+    }
+
+    fn post_load(context: C, values: &[Option<Vec<u8>>]) -> Result<Self, ViewError> {
+        let value = values.first().ok_or(ViewError::PostLoadValuesError)?;
+        let size_and_state = from_bytes_option::<SizeAndState, _>(value)?;
+        let is_small = match size_and_state {
+            None => true,
+            Some(SizeAndState {
+                total_size: _,
+                state,
+            }) => state.is_some(),
+        };
+        Ok(match is_small {
+            true => Self::Small(SmallKeyValueStoreView::post_load(context, values)?),
+            false => Self::Big(BigKeyValueStoreView::post_load(context, values)?),
+        })
```

Comment:
Please merge the two `match`s and remove `is_small`.

Reply:
Ok.